autotelic/ietf-hmac-signatures
# Feat/sign method

### Summary
- Created new `createSignRequest` function to return a method that will add signature headers to a req object
  - Uses a modified version of `constructSignatureString` method from `fastify-hmac`
- Added `constructDigestString` method to calculate a digest header value based on a Request object body.
  - Modified from`getDigest` in `fastify-hmac`
- Added modified version of `getAlgorithm` from `fastify-hmac`
- Added tests
- Added method and usage documentation

### Example scripts (and dependencies)
- The example scripts and dependencies were removed from this PR in [this commit](https://github.com/autotelic/ietf-hmac-signatures/pull/5/commits/e875f0e07cacacebabff12a58ec372addc8ea560) as these examples and dependencies should not end up in the package release. 
- In order to test this method locally, you will either need to create your own test script and run a local instance of `fastify-hmac` or simpy temporarily revert the commit where I removed the example and follow the test plan below.
  - basic 'client' script to create and send requests to a server for development and testing
    - script uses `got` api, `cross-fetch` api, axios api and HTTP api to test `signRequest` method use
  - Basic example server from `fastify-hmac` to receive test requests from client script

### Test Plan
- clone and install `npm i`
- run dev Server using `fastify-hmac` for request verification `npm run example:basicServer`
- Create, sign and send a 'client' fetch request to the dev server by executing `example/basicRequest.js` in a separate terminal - `node example/basicReqest.js`
- On a successful request signing and server verification: 
  - The server termial will show a log with `"res":{"statusCode":200}`
  - The 'client' terminal will console log out the server response message `Response Message:  { foo: 'bar' }`
- On a failed request signing and server verification: 
  - The server termial will show a log with `"res":{"statusCode":401}`
  - The 'client' terminal will console log out the server response message `Response Message:  {
  statusCode: 401,
  error: 'Unauthorized',
  message: 'Signature verification failed'
}`

- Change request parameters in `example/basicRequest.js` and confirm requests are still verified by `fastify-hmac` example server
  - Changing and of the shared options between the server and the client will cause the verification to fail i.e. `sharedSecret`, `algorithmMap`, etc.
  - Changing the `signedHeaders`, Request body, etc. will result in a different signature that should still be verified by the example server using `fastify-hmac`.
